File: clients/ruby/test/wire_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "novamem"

class WireTest < Minitest::Test
  # proved by: skipping nil before checking `required` in to_h_wire fails this test.
  def test_required_nullable_fields_are_sent_as_null
    assert_equal({ "entries" => [], "nextAfterId" => nil },
                 Novamem::ExportPage.new(entries: [], next_after_id: nil).to_h_wire)
  end

  def test_unset_optional_fields_are_omitted
    assert_equal({ "query" => "coffee" }, Novamem::SearchRequest.new(query: "coffee", namespace: "").to_h_wire)
  end

  # proved by: passing strings through Wire.timestamp unchanged fails this test.
  def test_timestamps_are_sent_as_utc_with_a_z
    assert_equal({ "since" => "2026-09-24T10:00:00.000Z" },
                 Novamem::RecentRequest.new(since: "2026-09-24T12:00:00+02:00").to_h_wire)
    assert_equal({ "since" => "yesterday-ish" }, Novamem::RecentRequest.new(since: "yesterday-ish").to_h_wire)
  end

  def test_each_type_has_its_own_fields
    refute_equal Novamem::SearchRequest::FIELDS, Novamem::ForgetResult::FIELDS
  end
end

File: clients/php/tests/WireTest.php
<?php

declare(strict_types=1);

use Novamem\Types as T;
use PHPUnit\Framework\TestCase;

final class WireTest extends TestCase
{
    // proved by: encoding an empty map as array_map(...) instead of an
    // object fails this test (json_encode([]) is "[]").
    public function testAnEmptyMapIsSentAsAnObject(): void
    {
        $this->assertSame(
            '{"content":"c","metadata":{}}',
            json_encode(
                T\CaptureRequest::fromArray([
                    "content" => "c",
                    "metadata" => [],
                ])->toArray(),
            ),
        );
    }

    public function testRequiredNullableFieldsAreSentAsNull(): void
    {
        $this->assertSame(
            ["entries" => [], "nextAfterId" => null],
            new T\ExportPage(entries: [], nextAfterId: null)->toArray(),
        );
    }

    public function testUnsetOptionalFieldsAreOmitted(): void
    {
        $this->assertSame(
            ["query" => "coffee"],
            new T\SearchRequest(query: "coffee", namespace: "")->toArray(),
        );
    }

    // proved by: returning strings unchanged from Wire::timestamp fails this test.
    public function testTimestampsAreSentAsUtcWithAZ(): void
    {
        $this->assertSame(
            ["since" => "2026-09-24T10:00:00.000Z"],
            new T\RecentRequest(since: "2026-09-24T12:00:00+02:00")->toArray(),
        );
        $this->assertSame(
            ["since" => "yesterday-ish"],
            new T\RecentRequest(since: "yesterday-ish")->toArray(),
        );
    }

    public function testUnknownKeysAreIgnored(): void
    {
        $this->assertSame(
            ["content" => "c", "id" => "e1"],
            T\MemoryEntry::fromArray([
                "id" => "e1",
                "content" => "c",
                "future" => 1,
            ])->toArray(),
        );
    }
}
